File: golang/authentication/did_wba_authenticator.go
package authentication

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	anp "github.com/agent-network-protocol/anp/golang"
)

// DIDWbaAuthHeader manages outgoing DIDWba and HTTP signature authentication headers.
type DIDWbaAuthHeader struct {
	didDocumentPath string
	privateKeyPath  string
	authMode        AuthMode
	didDocument     map[string]any
	tokens          map[string]string
}

// NewDIDWbaAuthHeader creates a new auth header helper.
func NewDIDWbaAuthHeader(didDocumentPath string, privateKeyPath string, authMode AuthMode) *DIDWbaAuthHeader {
	if authMode == "" {
		authMode = AuthModeHTTPSignatures
	}
	return &DIDWbaAuthHeader{didDocumentPath: filepath.Clean(didDocumentPath), privateKeyPath: filepath.Clean(privateKeyPath), authMode: authMode, tokens: map[string]string{}}
}

// GetAuthHeader returns the next outbound authentication headers.
func (h *DIDWbaAuthHeader) GetAuthHeader(serverURL string, forceNew bool, method string, headers map[string]string, body []byte) (map[string]string, error) {
	domain := extractDomain(serverURL)
	if !forceNew {
		if token, ok := h.tokens[domain]; ok {
			return map[string]string{"Authorization": "Bearer " + token}, nil
		}
	}
	didDocument, err := h.loadDidDocument()
	if err != nil {
		return nil, err
	}
	privateKey, err := h.loadPrivateKey()
	if err != nil {
		return nil, err
	}
	switch h.authMode {
	case AuthModeLegacyDidWba:
		authorization, err := GenerateAuthHeader(didDocument, domain, privateKey, "1.1")
		if err != nil {
			return nil, err
		}
		return map[string]string{"Authorization": authorization}, nil
	case AuthModeAuto, AuthModeHTTPSignatures:
		return GenerateHTTPSignatureHeaders(didDocument, serverURL, method, privateKey, headers, body, HttpSignatureOptions{})
	default:
		return nil, errUnsupportedAuthMode(h.authMode)
	}
}

// UpdateToken stores a bearer token discovered in response headers.
func (h *DIDWbaAuthHeader) UpdateToken(serverURL string, headers map[string]string) string {
	domain := extractDomain(serverURL)
	if value, ok := getHeaderCaseInsensitive(headers, "Authentication-Info"); ok {
		parsed := parseAuthenticationInfo(value)
		if token := parsed["access_token"]; token != "" {
			h.tokens[domain] = token
			return token
		}
	}
	if value, ok := getHeaderCaseInsensitive(headers, "Authorization"); ok && strings.HasPrefix(value, "Bearer ") {
		token := strings.TrimPrefix(value, "Bearer ")
		h.tokens[domain] = token
		return token
	}
	return ""
}

// ClearToken removes a cached bearer token for the given server.
func (h *DIDWbaAuthHeader) ClearToken(serverURL string) {
	delete(h.tokens, extractDomain(serverURL))
}

// ClearAllTokens clears all cached bearer tokens.
func (h *DIDWbaAuthHeader) ClearAllTokens() {
	h.tokens = map[string]string{}
}

// ShouldRetryAfter401 determines whether a 401 response should be retried with a new challenge response.
func (h *DIDWbaAuthHeader) ShouldRetryAfter401(responseHeaders map[string]string) bool {
	wwwAuthenticate, ok := getHeaderCaseInsensitive(responseHeaders, "WWW-Authenticate")
	if !ok {
		return false
	}
	challenge := parseWWWAuthenticate(wwwAuthenticate)
	if challenge["nonce"] != "" {
		return true
	}
	errorValue := challenge["error"]
	return errorValue != "invalid_did" && errorValue != "invalid_verification_method" && errorValue != "forbidden_did"
}

// GetChallengeAuthHeader returns challenge-specific authentication headers.
func (h *DIDWbaAuthHeader) GetChallengeAuthHeader(serverURL string, responseHeaders map[string]string, method string, headers map[string]string, body []byte) (map[string]string, error) {
	challenge := map[string]string{}
	if value, ok := getHeaderCaseInsensitive(responseHeaders, "WWW-Authenticate"); ok {
		challenge = parseWWWAuthenticate(value)
	}
	var covered []string
	if value, ok := getHeaderCaseInsensitive(responseHeaders, "Accept-Signature"); ok {
		covered = normalizeCoveredComponents(parseAcceptSignature(value), headers, body)
	}
	didDocument, err := h.loadDidDocument()
	if err != nil {
		return nil, err
	}
	privateKey, err := h.loadPrivateKey()
	if err != nil {
		return nil, err
	}
	switch h.authMode {
	case AuthModeLegacyDidWba:
		authorization, err := generateAuthHeaderWithOverrides(didDocument, extractDomain(serverURL), privateKey, "1.1", challenge["nonce"], "")
		if err != nil {
			return nil, err
		}
		return map[string]string{"Authorization": authorization}, nil
	case AuthModeAuto, AuthModeHTTPSignatures:
		return GenerateHTTPSignatureHeaders(didDocument, serverURL, method, privateKey, headers, body, HttpSignatureOptions{Nonce: challenge["nonce"], CoveredComponents: covered})
	default:
		return nil, errUnsupportedAuthMode(h.authMode)
	}
}

func (h *DIDWbaAuthHeader) loadDidDocument() (map[string]any, error) {
	if h.didDocument != nil {
		return h.didDocument, nil
	}
	data, err := os.ReadFile(h.didDocumentPath)
	if err != nil {
		return nil, err
	}
	var document map[string]any
	if err := decodeJSON(strings.NewReader(string(data)), &document); err != nil {
		return nil, err
	}
	h.didDocument = document
	return document, nil
}

func (h *DIDWbaAuthHeader) loadPrivateKey() (anp.PrivateKeyMaterial, error) {
	data, err := os.ReadFile(h.privateKeyPath)
	if err != nil {
		return anp.PrivateKeyMaterial{}, err
	}
	return anp.PrivateKeyFromPEM(string(data))
}

func parseAuthenticationInfo(value string) map[string]string {
	return parseDelimitedParams(value, ',')
}

func parseWWWAuthenticate(value string) map[string]string {
	trimmed := strings.TrimSpace(value)
	trimmed = strings.TrimPrefix(strings.TrimPrefix(trimmed, "DIDWba "), "didwba ")
	return parseDelimitedParams(trimmed, ',')
}

func parseAcceptSignature(value string) []string {
	matches := acceptSignatureRegexp.FindAllStringSubmatch(value, -1)
	components := make([]string, 0, len(matches))
	for _, match := range matches {
		if len(match) == 2 {
			components = append(components, match[1])
		}
	}
	return components
}

func normalizeCoveredComponents(components []string, headers map[string]string, body []byte) []string {
	if len(components) == 0 {
		return nil
	}
	normalizedHeaders := map[string]string{}
	for key, value := range headers {
		normalizedHeaders[strings.ToLower(key)] = value
	}
	bodyPresent := len(body) > 0
	result := make([]string, 0, len(components))
	for _, component := range components {
		lower := strings.ToLower(component)
		switch {
		case lower == "content-digest" && !bodyPresent:
			continue
		case lower == "content-length" && !bodyPresent && normalizedHeaders["content-length"] == "":
			continue
		case lower == "content-type" && normalizedHeaders["content-type"] == "":
			continue
		case !strings.HasPrefix(lower, "@") && lower != "content-length" && lower != "content-digest" && normalizedHeaders[lower] == "":
			continue
		default:
			result = append(result, component)
		}
	}
	return result
}

func parseDelimitedParams(value string, separator rune) map[string]string {
	result := map[string]string{}
	for _, part := range strings.Split(value, string(separator)) {
		trimmed := strings.TrimSpace(part)
		if trimmed == "" {
			continue
		}
		name, rawValue, ok := strings.Cut(trimmed, "=")
		if !ok {
			continue
		}
		result[strings.TrimSpace(name)] = strings.Trim(strings.TrimSpace(rawValue), "\"")
	}
	return result
}

func errUnsupportedAuthMode(mode AuthMode) error {
	return fmt.Errorf("unsupported auth mode: %s", mode)
}

var acceptSignatureRegexp = regexp.MustCompile(`"([^"]+)"`)

File: rust/src/authentication/did_wba_authenticator.rs
use std::collections::BTreeMap;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use regex::Regex;
use serde_json::Value;
#[cfg(feature = "network")]
use url::Url;

use crate::keys::PrivateKeyMaterial;

use super::did_wba::{
    generate_auth_header, generate_auth_header_with_overrides, AuthenticationError,
};
use super::http_signatures::{generate_http_signature_headers, HttpSignatureOptions};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AuthMode {
    HttpSignatures,
    LegacyDidWba,
    Auto,
}

impl AuthMode {
    pub fn from_str(value: &str) -> Self {
        match value.to_ascii_lowercase().as_str() {
            "legacy_didwba" => Self::LegacyDidWba,
            "auto" => Self::Auto,
            _ => Self::HttpSignatures,
        }
    }
}

#[derive(Debug)]
pub struct DIDWbaAuthHeader {
    did_document_path: PathBuf,
    private_key_path: PathBuf,
    auth_mode: AuthMode,
    did_document_cache: Option<Value>,
    tokens: HashMap<String, String>,
}

impl DIDWbaAuthHeader {
    pub fn new(
        did_document_path: impl AsRef<Path>,
        private_key_path: impl AsRef<Path>,
        auth_mode: AuthMode,
    ) -> Self {
        Self {
            did_document_path: did_document_path.as_ref().to_path_buf(),
            private_key_path: private_key_path.as_ref().to_path_buf(),
            auth_mode,
            did_document_cache: None,
            tokens: HashMap::new(),
        }
    }

    pub fn get_auth_header(
        &mut self,
        server_url: &str,
        force_new: bool,
        method: &str,
        headers: Option<&std::collections::BTreeMap<String, String>>,
        body: Option<&[u8]>,
    ) -> Result<std::collections::BTreeMap<String, String>, AuthenticationError> {
        let domain = extract_domain(server_url);
        if !force_new {
            if let Some(token) = self.tokens.get(&domain) {
                let mut result = std::collections::BTreeMap::new();
                result.insert("Authorization".to_string(), format!("Bearer {}", token));
                return Ok(result);
            }
        }

        let did_document = self.load_did_document()?.clone();
        let private_key = self.load_private_key()?;
        match self.auth_mode {
            AuthMode::HttpSignatures | AuthMode::Auto => generate_http_signature_headers(
                &did_document,
                server_url,
                method,
                &private_key,
                headers,
                body,
                HttpSignatureOptions::default(),
            )
            .map_err(|_| AuthenticationError::SignatureGenerationFailed),
            AuthMode::LegacyDidWba => {
                let value = generate_auth_header(&did_document, &domain, &private_key, "1.1")?;
                let mut result = std::collections::BTreeMap::new();
                result.insert("Authorization".to_string(), value);
                Ok(result)
            }
        }
    }

    pub fn update_token(
        &mut self,
        server_url: &str,
        headers: &std::collections::BTreeMap<String, String>,
    ) -> Option<String> {
        let domain = extract_domain(server_url);
        if let Some(value) = get_header_case_insensitive(headers, "Authentication-Info") {
            let parsed = parse_authentication_info(value);
            if let Some(token) = parsed.get("access_token").filter(|token| !token.is_empty()) {
                self.tokens.insert(domain, token.clone());
                return Some(token.clone());
            }
        }
        if let Some(value) = get_header_case_insensitive(headers, "Authorization") {
            if let Some(token) = value.strip_prefix("Bearer ") {
                let token = token.to_string();
                self.tokens.insert(domain, token.clone());
                return Some(token);
            }
        }
        None
    }

    pub fn clear_token(&mut self, server_url: &str) {
        let domain = extract_domain(server_url);
        self.tokens.remove(&domain);
    }

    pub fn clear_all_tokens(&mut self) {
        self.tokens.clear();
    }

    pub fn should_retry_after_401(&self, response_headers: &BTreeMap<String, String>) -> bool {
        let Some(www_authenticate) =
            get_header_case_insensitive(response_headers, "WWW-Authenticate")
        else {
            return false;
        };
        let challenge = parse_www_authenticate(www_authenticate);
        if challenge.get("nonce").is_some() {
            return true;
        }
        !matches!(
            challenge.get("error").map(|value| value.as_str()),
            Some("invalid_did") | Some("invalid_verification_method") | Some("forbidden_did")
        )
    }

    pub fn get_challenge_auth_header(
        &mut self,
        server_url: &str,
        response_headers: &BTreeMap<String, String>,
        method: &str,
        headers: Option<&BTreeMap<String, String>>,
        body: Option<&[u8]>,
    ) -> Result<BTreeMap<String, String>, AuthenticationError> {
        let www_authenticate = get_header_case_insensitive(response_headers, "WWW-Authenticate");
        let accept_signature = get_header_case_insensitive(response_headers, "Accept-Signature");
        let challenge = www_authenticate
            .map(|value| parse_www_authenticate(value))
            .unwrap_or_default();
        let covered_components = normalize_covered_components(
            accept_signature
                .map(|value| parse_accept_signature(value))
                .as_ref(),
            headers,
            body,
        );
        let nonce = challenge.get("nonce").map(String::as_str);

        let did_document = self.load_did_document()?.clone();
        let private_key = self.load_private_key()?;
        match self.auth_mode {
            AuthMode::HttpSignatures | AuthMode::Auto => generate_http_signature_headers(
                &did_document,
                server_url,
                method,
                &private_key,
                headers,
                body,
                HttpSignatureOptions {
                    nonce: nonce.map(ToOwned::to_owned),
                    covered_components,
                    ..HttpSignatureOptions::default()
                },
            )
            .map_err(|_| AuthenticationError::SignatureGenerationFailed),
            AuthMode::LegacyDidWba => {
                let value = generate_auth_header_with_overrides(
                    &did_document,
                    &extract_domain(server_url),
                    &private_key,
                    "1.1",
                    nonce,
                    None,
                )?;
                let mut result = BTreeMap::new();
                result.insert("Authorization".to_string(), value);
                Ok(result)
            }
        }
    }

    fn load_did_document(&mut self) -> Result<&Value, AuthenticationError> {
        if self.did_document_cache.is_none() {
            let content = fs::read_to_string(&self.did_document_path)
                .map_err(|_| AuthenticationError::IoFailure)?;
            let value =
                serde_json::from_str(&content).map_err(|_| AuthenticationError::JsonFailure)?;
            self.did_document_cache = Some(value);
        }
        self.did_document_cache
            .as_ref()
            .ok_or(AuthenticationError::InvalidDidDocument)
    }

    fn load_private_key(&self) -> Result<PrivateKeyMaterial, AuthenticationError> {
        let content = fs::read_to_string(&self.private_key_path)
            .map_err(|_| AuthenticationError::IoFailure)?;
        PrivateKeyMaterial::from_pem(&content).map_err(|_| AuthenticationError::InvalidDidDocument)
    }
}

fn extract_domain(server_url: &str) -> String {
    #[cfg(feature = "network")]
    {
        return Url::parse(server_url)
            .ok()
            .and_then(|value| value.host_str().map(|host| host.to_string()))
            .unwrap_or_else(|| server_url.to_string());
    }

    #[cfg(not(feature = "network"))]
    {
        server_url
            .split_once("://")
            .map(|(_, rest)| rest)
            .and_then(|rest| rest.split(['/', '?', '#']).next())
            .and_then(|authority| {
                authority
                    .rsplit_once('@')
                    .map(|(_, host)| host)
                    .or(Some(authority))
            })
            .map(|authority| {
                if let Some(stripped) = authority.strip_prefix('[') {
                    stripped
                        .split_once(']')
                        .map(|(host, _)| host.to_string())
                        .unwrap_or_else(|| authority.to_string())
                } else {
                    authority
                        .split_once(':')
                        .map(|(host, _)| host.to_string())
                        .unwrap_or_else(|| authority.to_string())
                }
            })
            .filter(|host| !host.is_empty())
            .unwrap_or_else(|| server_url.to_string())
    }
}

fn get_header_case_insensitive<'a>(
    headers: &'a std::collections::BTreeMap<String, String>,
    name: &str,
) -> Option<&'a String> {
    headers
        .iter()
        .find(|(key, _)| key.eq_ignore_ascii_case(name))
        .map(|(_, value)| value)
}

fn parse_authentication_info(value: &str) -> HashMap<String, String> {
    value
        .split(',')
        .filter_map(|item| item.trim().split_once('='))
        .map(|(key, raw)| {
            (
                key.trim().to_string(),
                raw.trim().trim_matches('"').to_string(),
            )
        })
        .collect()
}

fn parse_www_authenticate(value: &str) -> HashMap<String, String> {
    let normalized = value
        .trim()
        .strip_prefix("DIDWba ")
        .or_else(|| value.trim().strip_prefix("didwba "))
        .unwrap_or(value.trim());
    let regex = Regex::new(r#"([\w-]+)=("[^"]*"|[^,]+)"#).expect("regex should compile");
    regex
        .captures_iter(normalized)
        .filter_map(|captures| {
            let key = captures.get(1)?.as_str().trim().to_string();
            let value = captures
                .get(2)?
                .as_str()
                .trim()
                .trim_matches('"')
                .to_string();
            Some((key, value))
        })
        .collect()
}

fn parse_accept_signature(value: &str) -> Vec<String> {
    let regex = Regex::new(r#""([^"]+)""#).expect("regex should compile");
    regex
        .captures_iter(value)
        .filter_map(|captures| captures.get(1).map(|matched| matched.as_str().to_string()))
        .collect()
}

fn normalize_covered_components(
    covered_components: Option<&Vec<String>>,
    headers: Option<&BTreeMap<String, String>>,
    body: Option<&[u8]>,
) -> Option<Vec<String>> {
    let covered_components = covered_components?;
    let body_present = body.map(|bytes| !bytes.is_empty()).unwrap_or(false);
    let normalized_headers = headers
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .filter_map(|(key, value)| (!value.is_empty()).then(|| (key.to_ascii_lowercase(), value)))
        .collect::<BTreeMap<_, _>>();

    let mut result = Vec::new();
    for component in covered_components {
        let lower = component.to_ascii_lowercase();
        if lower == "content-digest" && !body_present {
            continue;
        }
        if lower == "content-length"
            && !body_present
            && !normalized_headers.contains_key("content-length")
        {
            continue;
        }
        if lower == "content-type" && !normalized_headers.contains_key("content-type") {
            continue;
        }
        if !lower.starts_with('@')
            && lower != "content-length"
            && lower != "content-digest"
            && !normalized_headers.contains_key(&lower)
        {
            continue;
        }
        result.push(component.clone());
    }
    (!result.is_empty()).then_some(result)
}
